### PR TITLE
Fix draw order of transparent materials with multiple directional lights

### LINKED ISSUE #47078: Shaded Transparent Material Is Drawn On Top Of Unshaded Transparent Material With Two Directional Lights
**Godot version:**
v3.2.4-rc5

**OS/device including version:**
Windows 64
GLES 3
GeForce GTX 760

**Issue description:**
Transparent materials with shading enabled (no unshaded) are drawn using the additive blend mode, even if the blend mode is set to blend_mix.
This only occurs in the presence of more than one directional light. (With spot lights, this does not happen.)
This seems to be the way directional lights are implemented, i.e. they are rendered in multiple passes.

This wouldn't be so bad, but they are also drawn AFTER (i.e. on top of) other transparent materials, even if those are unshaded and have a higher priority.
I would expect the unshaded material (or any transparent material with higher priority) to be drawn afterwards, which would avoid this issue.

_A bit of background:
This happened to me when I was using multiple particle systems.
One for clouds, which are affected by light, and
another particle system, with small mostly opaque particles, which should be drawn in front._

**Steps to reproduce:**
Create two objects, with a ShaderMaterial each.
Add `ALPHA=1.0;` to both, in order to make them transparent materials.
Add `render_mode unshaded;` to one of them.
Add two directional lights. 

**Minimal reproduction project:**
[draw_order_bug.zip](https://github.com/godotengine/godot/files/6152335/draw_order_bug.zip)

As you can see, the green plane is visible through the red planes.
If you disable one of the directional lights, the problem goes away.
Changing priority does not affect it.

### PR DESCRIPTION
**Pull request by mortarroad**
Fixes #47078

### CODE REVIEW

**Comment by mortarroad** [comment]
Never mind, this doesn't seem to work in all cases yet.

**Comment by mortarroad** [comment]
OK, now it also respects the material priority among multiple transparent shaded materials.
I'm not sure if we really want to support that too; I was mostly concerned about unshaded materials being drawn below shaded ones.
Though it is pretty useful to avoid flickering due to sorting.

**Review by clayjohn — COMMENTED** [review]
On a preliminary review, this looks quite good. I need to spend more time reviewing before I approve though.

I have a few performance concerns in the back of my mind, but I think overall it shouldn't be too bad as not many people use multiple directional lights.

**Comment by mortarroad** [comment]
Another related issue, that does not get fixed by this (i.e. it did not work before, nor does it now):
If a MeshInstance is on a layer with a directional light, but not on the layer with the first directional light in the scene, it will only be drawn in the additive pass.

That means you can't have a scene with two transparent meshes and two lights, with the first mesh being lit by the first light and the second mesh lit by the second light.

Fixing that would require grouping the elements by lights...
We should definitely document that there are issues with transparent objects being lit by directional lights.

**Review by JFonS — APPROVED** [review]
The code makes sense. I did some testing and everything seems to work as expected :)

I don't know if it's worth merging for 3.3 though. On one side, if we are fixing multiple directional lights in 3.3 we might as well do the best we can. On the other, we are introducing more chances for regressions in a late release candidate, for something that is not very widely used...

**tl;dr**
Looks great, we need to decide if we merge before or after 3.3 :)

**Comment by akien-mga** [comment]
I think it's better to play it safe and merge for 3.4. The two commits should be squashed into one I guess (or if the second one makes sense as standalone, it should have a clearer title).

**Comment by mortarroad** [comment]
Squashed and rebased to latest 3.x.

**Comment by akien-mga** [comment]
Thanks!
